node/src/syncer.rs

```diff
@@ -1,72 +1,634 @@
+                }
+            }
+        }
     }
 
-    async fn send_command(&self, _cmd: SyncerCmd) -> Result<()> {
-        // TODO
-        Ok(())
+    async fn syncing_info(&self) -> SyncingInfo {
+        SyncingInfo {
+            local_head: self.store.head_height().await.unwrap_or(0),
+            subjective_head: self.subjective_head_height.unwrap_or(0),
+        }
+    }
+
+    #[instrument(skip_all)]
+    async fn report(&mut self) {
+        let SyncingInfo {
+            local_head,
+            subjective_head,
+        } = self.syncing_info().await;
+
+        let ongoing_batch = self
+            .ongoing_batch
+            .map(|(start, end)| format!("[{start}, {end}]"))
+            .unwrap_or_else(|| "None".to_string());
+
+        info!("syncing: {local_head}/{subjective_head}, ongoing batch: {ongoing_batch}",);
+    }
+
+    fn spawn_try_init(&self) -> oneshot::Receiver<u64> {
+        let p2p = self.p2p.clone();
+        let store = self.store.clone();
+        let genesis_hash = self.genesis_hash;
+        let (tx, rx) = oneshot::channel();
+
+        spawn(
+            async move {
+                loop {
+                    match try_init(&p2p, &store, genesis_hash).await {
+                        Ok(height) => {
+                            tx.maybe_send(height);
+                            return;
+                        }
+                        Err(e) => {
+                            warn!("Intialization of subjective head failed: {e}. Trying again.")
+                        }
+                    }
+                }
+            }
+            .instrument(info_span!("try_init")),
+        );
+
+        rx
+    }
+
+    async fn on_cmd(&mut self, cmd: SyncerCmd) {
+        match cmd {
+            SyncerCmd::GetInfo { respond_to } => {
+                let info = self.syncing_info().await;
+                respond_to.maybe_send(info);
+            }
+        }
+    }
+
+    #[instrument(skip_all)]
+    async fn on_header_sub_message(&mut self) {
+        // If subjective head then do nothing.
+        // We do this to avoid some edge cases.
+        if self.subjective_head_height.is_none() {
+            return;
+        }
+
+        let Some(new_head) = self.header_sub_watcher.borrow().to_owned() else {
+            // Nothing to do
+            return;
+        };
+
+        let new_head_height = new_head.height().value();
+
+        // We don't want to interfere with any ongoing batch fetching
+        if self.ongoing_batch.is_none() {
+            if let Ok(store_head_height) = self.store.head_height().await {
+                // If our new header is adjacent to the HEAD of the store
+                if store_head_height + 1 == new_head_height {
+                    // Header is already verified by HeaderSub
+                    if self.store.append_single_unchecked(new_head).await.is_ok() {
+                        info!("Added {new_head_height} from HeaderSub");
```

Comment:
```suggestion
                        info!("Added header {new_head_height} from HeaderSub");
```
a nit but I think it'll look better in logs